### PR TITLE
Add Camera WASD controls plugin (WIP)

### PR DESCRIPTION
**Pull request by ChaoticConundrum**
As the title says. This addresses #1008. This is **NOT** ready to merge yet.
Pressing Enter toggles remapping WASD to arrow keys / suppressing typing events in the chat box.

A fixed tooltip is currently drawn on the chat box when chat entry is disabled:
screenshot_20180320_015510
This is messy, because it also has to be conditionally hidden when the chat box is hidden/covered (todo).

It may be possible to add text to the entry line (like RS2 did), but we would need to be sure it would not be accidentally sent.

### CODE REVIEW

**Comment by ChaoticConundrum** [comment]
V2 of this changes the widget text on the chat entry line when WASD is remapped:
screenshot_20180321_001103

The nice thing about this is the chat text buffer is unmodified, so the user can toggle in and out of chat mode without losing their message. The downside of this is keeping up with the widget being repopulated by literally any keyboard input to the runescape client.

File: runelite-client/src/main/java/net/runelite/client/ui/overlay/tooltip/TooltipOverlay.java [hunk 1]
```diff
@@ -82,31 +82,34 @@ public Dimension render(Graphics2D graphics, Point parent)
 					: new net.runelite.api.Point(0, 0);
 
 				position.setLocation(mouseCanvasPosition.getX(), mouseCanvasPosition.getY());
+
+				// check if this tooltip would overlap the last
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/ui/overlay/tooltip/TooltipOverlay.java]: As you are not using tooltips in the PR anymore, revert this commit as it is out of scope of the PR

File: runelite-client/src/main/java/net/runelite/client/plugins/wasdcamera/WASDCameraInputListener.java [hunk 2]
```diff
@@ -52,24 +52,9 @@ public void keyPressed(KeyEvent e)
...
-			{
-				case KeyEvent.VK_ESCAPE:
-					e.consume();
-				case KeyEvent.VK_ENTER:
-					plugin.setChatMode(false);
-					break;
-			}
-		}
-		else
+		if (plugin.shouldRemap())
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/wasdcamera/WASDCameraInputListener.java]: Merge this commit with previous commit

Comment (ChaoticConundrum) [reply runelite-client/src/main/java/net/runelite/client/plugins/wasdcamera/WASDCameraInputListener.java]: I will rebase the PR when everything works

**Comment by Mamboosun** [comment]
Is this in working order?

**Comment by ChaoticConundrum** [comment]
No, I'm closing this pull request. I may try this again in the future, but it requires some client API work to work properly.

The problem I wasn't able to solve was dialogs and text entry widgets would not work while the WASD camera controls were enabled without pressing enter first, because the key events were intercepted when the camera control were enabled. I tried detecting when widgets that used the keyboard were visible, which mostly worked, but it seemed like a bad solution in the long run.
